File: libraries/dotnet/UrcConverter/Sources/Ojn/Parse.fs
namespace UrcConverter.Sources.Ojn

module Parse =

    open System
    open System.Buffers.Binary
    open System.Text
    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources
    open UrcConverter.Sources.Ojn.Model

    let private stringRanges = [ (108, 172); (172, 204); (204, 236); (236, 268) ]

    let private signature = [| 0x6fuy; 0x6auy; 0x6euy; 0uy |]

    let private readI32 (data: byte[]) (offset: int) : int =
        BinaryPrimitives.ReadInt32LittleEndian(ReadOnlySpan<byte>(data, offset, 4))

    let private readU16 (data: byte[]) (offset: int) : uint16 =
        BinaryPrimitives.ReadUInt16LittleEndian(ReadOnlySpan<byte>(data, offset, 2))

    let private readI16 (data: byte[]) (offset: int) : int16 =
        BinaryPrimitives.ReadInt16LittleEndian(ReadOnlySpan<byte>(data, offset, 2))

    let private readF32 (data: byte[]) (offset: int) : float =
        float (BinaryPrimitives.ReadSingleLittleEndian(ReadOnlySpan<byte>(data, offset, 4)))

    let private decrypt (data: byte[]) : byte[] =
        let block = int data[3]

        if block = 0 then
            data
        else
            let key = Array.create block data[4]
            key[0] <- data[6]
            key[block / 2] <- data[5]
            let size = data.Length
            Array.init (size - 8) (fun i -> data[size - 1 - i] ^^^ key[i % block])

    let private decodeStrings (data: byte[]) : Result<string * string * string, UrcError> =
        result {
            let fields =
                stringRanges
                |> List.map (fun (start, end') ->
                    let field = data[start .. end' - 1]

                    match field |> Array.tryFindIndex (fun byte -> byte = 0uy) with
                    | Some nul -> field[.. nul - 1]
                    | None -> field)

            let blob = Array.concat fields

            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
            let utf8 = UTF8Encoding(false, true)
            let cp949 = Encoding.GetEncoding(949, EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback)

            let validates (encoding: Encoding) (bytes: byte[]) =
                try
                    encoding.GetString bytes |> ignore
                    true
                with :? DecoderFallbackException ->
                    false

            let charset: Encoding option =
                if blob |> Array.forall (fun byte -> byte < 0x80uy) then Some utf8
                elif validates utf8 blob then Some utf8
                elif validates cp949 blob then Some cp949
                else None

            match charset with
            | None -> return! Error(UrcError.syntax 108 "strings are neither ASCII, UTF-8, nor CP949")
            | Some encoding ->
                let decoded =
                    fields
                    |> List.map (fun field ->
                        try
                            Ok(encoding.GetString field)
                        with :? DecoderFallbackException ->
                            Error(UrcError.syntax 108 "invalid string byte sequence"))

                let! strings = Shared.sequence decoded
                return strings[0], strings[1], strings[2]
        }

    let rec private readEvents
        (data: byte[])
        (offset: int)
        (bound: int)
        (channel: int)
        (measure: int)
        (total: int)
        (eventIndex: int)
        (acc: OjnEvent list)
        : Result<int * OjnEvent list, UrcError> =
        if eventIndex >= total then
            Ok(offset, List.rev acc)
        elif offset + 4 > bound then
            Error(UrcError.syntax offset "note section truncated")
        else
            let position = float eventIndex / float total

            let event kind value =
                { Measure = measure; Position = position; Channel = channel; Value = value; Kind = kind; Offset = offset }

            let acc' =
                if channel <= 1 then
                    let value = readF32 data offset
                    if value <= 0.0 then acc else event 0 value :: acc
                else
                    let value = readI16 data offset
                    let noteType = int data[offset + 3]

                    if channel >= 9 || value = 0s || noteType % 4 = 1 then
                        acc
                    else
                        event (noteType % 4) 0.0 :: acc

            readEvents data (offset + 4) bound channel measure total (eventIndex + 1) acc'

    let rec private readPackages
        (data: byte[])
        (offset: int)
        (bound: int)
        (count: int)
        (packageIndex: int)
        (acc: OjnEvent list list)
        : Result<OjnEvent list, UrcError> =
        if packageIndex >= count then
            Ok(acc |> List.rev |> List.concat)
        elif offset + 8 > bound then
            Error(UrcError.syntax offset "note section truncated")
        else
            let measure = readI32 data offset
            let channel = int (readU16 data (offset + 4))
            let total = int (readI16 data (offset + 6))

            if measure < 0 then
                Error(UrcError.syntax offset "negative measure index")
            else
                match readEvents data (offset + 8) bound channel measure (max total 0) 0 [] with
                | Error error -> Error error
                | Ok(nextOffset, acc') -> readPackages data nextOffset bound count (packageIndex + 1) (acc' :: acc)

    let private readDifficulty
        (data: byte[])
        (index: int)
        (start: int)
        (bound: int)
        (count: int)
        : Result<OjnDifficulty, UrcError> =
        result {
            if start < 300 then
                return! Error(UrcError.syntax start "note section overlaps the header")
            else
                let! events = readPackages data start bound count 0 []
                return { Index = index; Events = events }
        }

    /// Parses a plain or encrypted ("new" magic) OJN file into its source model.
    let parseOjn (bytes: byte[]) : Result<OjnFile, UrcError> =
        result {
            let data =
                if bytes.Length >= 8 && bytes[0] = 0x6euy && bytes[1] = 0x65uy && bytes[2] = 0x77uy then
                    decrypt bytes
                else
                    bytes

            if data.Length < 300 then
                return! Error(UrcError.syntax data.Length "file too short for OJN header")
            elif data[4..7] <> signature then
                return! Error(UrcError.syntax 4 "invalid OJN signature")
            else
                let bpm = readF32 data 16

                if bpm <= 0.0 then
                    return! Error(UrcError.syntax 16 "header BPM must be positive")
                else
                    let! title, artist, noter = decodeStrings data
                    let packageCounts = [| readI32 data 64; readI32 data 68; readI32 data 72 |]
                    let noteOffsets = [| readI32 data 284; readI32 data 288; readI32 data 292 |]
                    let coverOffset = readI32 data 296

                    let! difficulties =
                        [ 0 .. 2 ]
                        |> List.filter (fun index -> packageCounts[index] <> 0)
                        |> List.map (fun index ->
                            let endRaw = if index < 2 then noteOffsets[index + 1] else coverOffset
                            let endBound = min (max endRaw 0) data.Length
                            readDifficulty data index noteOffsets[index] endBound packageCounts[index])
                        |> Shared.sequence

                    if difficulties.IsEmpty then
                        return! Error(UrcError.syntax 0 "no difficulty with note packages")
                    else
                        return
                            {
                                Title = title
                                Artist = artist
                                Noter = noter
                                Bpm = bpm
                                Difficulties = difficulties
                            }
        }

File: libraries/python/src/urc_converter/sources/ojn/parse.py
"""Parser for O2Jam (.ojn) chart binaries."""

import struct

from ...error import UrcError
from .model import OjnDifficulty, OjnEvent, OjnFile

_STRING_FIELDS = ((108, 172), (172, 204), (204, 236), (236, 268))


def parse_ojn(data: bytes) -> OjnFile:
	"""Parse a plain or encrypted (``new`` magic) OJN file."""
	if data[:3] == b"new" and len(data) >= 8:
		data = _decrypt(data)
	if len(data) < 300:
		raise UrcError("syntax", len(data), "file too short for OJN header")
	if data[4:8] != b"ojn\x00":
		raise UrcError("syntax", 4, "invalid OJN signature")
	bpm = struct.unpack_from("<f", data, 16)[0]
	if bpm <= 0:
		raise UrcError("syntax", 16, "header BPM must be positive")

	title, artist, noter, _ojm = _decode_strings(data)
	package_counts = struct.unpack_from("<3i", data, 64)
	note_offsets = struct.unpack_from("<3i", data, 284)
	cover_offset = struct.unpack_from("<i", data, 296)[0]

	difficulties = []
	for index in range(3):
		if package_counts[index] == 0:
			continue
		start = note_offsets[index]
		end = note_offsets[index + 1] if index < 2 else cover_offset
		difficulties.append(
			_read_difficulty(data, index, start, min(end, len(data)), package_counts[index])
		)
	if not difficulties:
		raise UrcError("syntax", 0, "no difficulty with note packages")
	return OjnFile(title=title, artist=artist, noter=noter, bpm=bpm, difficulties=difficulties)


def _decrypt(data: bytes) -> bytes:
	block = data[3]
	if block == 0:
		return data
	key = [data[4]] * block
	key[0] = data[6]
	key[block // 2] = data[5]
	size = len(data)
	return bytes(data[size - 1 - i] ^ key[i % block] for i in range(size - 8))


def _read_difficulty(
	data: bytes, index: int, start: int, end: int, count: int
) -> OjnDifficulty:
	if start < 300:
		raise UrcError("syntax", start, "note section overlaps the header")
	events: list[OjnEvent] = []
	offset = start
	for _ in range(count):
		if offset + 8 > end:
			raise UrcError("syntax", offset, "note section truncated")
		measure, channel, total = struct.unpack_from("<iHh", data, offset)
		if measure < 0:
			raise UrcError("syntax", offset, "negative measure index")
		offset += 8
		for i in range(total):
			position = i / total
			if offset + 4 > end:
				raise UrcError("syntax", offset, "note section truncated")
			if channel in (0, 1):
				value = struct.unpack_from("<f", data, offset)[0]
				offset += 4
				if value <= 0:
					continue
				events.append(OjnEvent(measure, position, channel, value, 0, offset - 4))
			else:
				value, _volume_pan, type_ = struct.unpack_from("<hBB", data, offset)
				offset += 4
				if channel >= 9 or value == 0 or type_ % 4 == 1:
					continue
				events.append(OjnEvent(measure, position, channel, 0.0, type_ % 4, offset - 4))
	return OjnDifficulty(index=index, events=events)


def _decode_strings(data: bytes) -> tuple[str, str, str, str]:
	fields = [data[start:end].split(b"\x00", 1)[0] for start, end in _STRING_FIELDS]
	blob = b"".join(fields)
	if all(byte < 0x80 for byte in blob):
		charset = "ascii"
	elif _decodes(blob, "utf-8"):
		charset = "utf-8"
	elif _decodes(blob, "cp949"):
		charset = "cp949"
	else:
		raise UrcError("syntax", 108, "strings are neither ASCII, UTF-8, nor CP949")
	try:
		decoded = [field.decode(charset) for field in fields]
	except UnicodeDecodeError:
		raise UrcError("syntax", 108, "invalid string byte sequence") from None
	return decoded[0], decoded[1], decoded[2], decoded[3]


def _decodes(blob: bytes, charset: str) -> bool:
	try:
		blob.decode(charset)
	except UnicodeDecodeError:
		return False
	return True
